internal/users/localentries/localgroups.go

```diff
@@ -116,140 +226,142 @@ func existingLocalGroups(user, groupPath string) (groups []string, err error) {
+	return fmt.Sprintf("%s-", groupPath)
+}
 
-	// Get the list of local groups the user belong to
-	groups, err := existingLocalGroups(user, opts.groupPath)
-	if err != nil {
-		return err
-	}
+func formatGroupEntries(groups []types.GroupEntry) string {
+	groupLines := sliceutils.Map(groups, func(group types.GroupEntry) string {
+		return strings.Join([]string{
+			group.Name,
+			group.Passwd,
+			fmt.Sprint(group.GID),
+			strings.Join(group.Users, ","),
+		}, ":")
+	})
 
-	localGroupsMu.Lock()
-	defer localGroupsMu.Unlock()
-	for _, group := range groups {
-		args := opts.gpasswdCmd[1:]
-		args = append(args, "--delete", user, group)
-		if err := runGPasswd(opts.gpasswdCmd[0], args...); err != nil {
-			return err
-		}
-	}
+	// Add final new line to the group file.
+	groupLines = append(groupLines, "")
 
-	return nil
+	return strings.Join(groupLines, "\n")
 }
 
-// Clean removes all unexistent users from the local groups.
-func Clean(args ...Option) (err error) {
-	defer decorate.OnError(&err, "could not clean local groups completely")
+func (l *lockedGroups) saveLocalGroups(groups []types.GroupEntry) (err error) {
+	inputPath := l.options.groupInputPath
+	groupPath := l.options.groupOutputPath
 
-	opts := defaultOptions
-	for _, arg := range args {
-		arg(&opts)
-	}
+	defer decorate.OnError(&err, "could not write local groups to %q", groupPath)
 
-	localGroupsMu.Lock()
-	defer localGroupsMu.Unlock()
+	if slices.EqualFunc(l.currentGroups, groups, types.GroupEntry.Equals) {
+		log.Debugf(context.TODO(), "Nothing to do, groups are equal")
+		return nil
+	}
 
-	// Add the existingUsers to a map to speed up search
-	existingUsers := make(map[string]struct{})
-	usernames, err := opts.getUsersFunc()
-	if err != nil {
+	if err := types.ValidateGroupEntries(groups); err != nil {
 		return err
 	}
-	for _, username := range usernames {
-		existingUsers[username] = struct{}{}
-	}
-	// If no username was returned, something probably went wrong during the getpwent call and we should stop,
-	// otherwise we would remove all users from the local groups.
-	if len(existingUsers) == 0 {
-		return errors.New("no existing users found, local groups won't be cleaned")
-	}
 
-	// Get the list of local groups
-	f, err := os.Open(opts.groupPath)
-	if err != nil {
-		return err
+	backupPath := groupFileBackupPath(groupPath)
+	oldBackup := ""
+	groupsEntries := formatGroupEntries(groups)
+
+	log.Debugf(context.TODO(), "Saving group entries %#v to %q", groups, groupPath)
+	if len(groupsEntries) > 0 {
+		log.Debugf(context.TODO(), "Group file content:\n%s", groupsEntries)
 	}
-	defer f.Close()
 
-	// Format of a line composing the group file is:
-	// group_name:password:group_id:user1,…,usern
-	var delOps [][]string
-	scanner := bufio.NewScanner(f)
-	for scanner.Scan() {
-		t := strings.TrimSpace(scanner.Text())
-		if t == "" {
-			continue
+	if tmpDir, err := os.MkdirTemp(os.TempDir(), "authd-groups-backup"); err == nil {
```

Comment:
```suggestion

	// Try to create a backup of the backup file (in case it exists), so that we can move it
	// back in case the creation of the new backup fails.
	if tmpDir, err := os.MkdirTemp(os.TempDir(), "authd-groups-backup"); err == nil {
```